### PR TITLE
Catch FormatException caused by bad simctl output

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description
Sometimes simctl seems to return bad output. Not sure what the exact cause is because it seems fairly rare. Instead of crashing, we should return an empty map/no devices.

```
FormatException: Unexpected character (at character 1)
Install Started
^

  | at _ChunkedJsonParser.fail | (convert_patch.dart:1392 )
-- | -- | --
  | at _ChunkedJsonParser.parseNumber | (convert_patch.dart:1259 )
  | at _ChunkedJsonParser.parse | (convert_patch.dart:924 )
  | at _parseJson | (convert_patch.dart:29 )
  | at JsonDecoder.convert | (json.dart:493 )
  | at JsonCodec.decode | (json.dart:151 )
  | at SimControl._list | (simulators.dart:92 )
  | at SimControl.getDevices | (simulators.dart:99 )
  | at SimControl.getConnectedDevices | (simulators.dart:113 )
  | at IOSSimulatorUtils.getAttachedDevices | (simulators.dart:53 )
  | at IOSSimulators.pollingGetDevices | (simulators.dart:42 )
  | at <asynchronous gap> | (async )
  | at PollingDeviceDiscovery.devices | (device.dart:270 )
  | at <asynchronous gap> | (async )
  | at DeviceManager.getAllConnectedDevices | (device.dart:136 )
  | at <asynchronous gap> | (async )
  | at DeviceManager.getDevicesById | (device.dart:101 )
  | at <asynchronous gap> | (async )
  | at DeviceManager.getDevices | (device.dart:125 )
  | at DeviceManager.findTargetDevices | (device.dart:169 )
  | at <asynchronous gap> | (async )
  | at FlutterCommand.findAllTargetDevices | (flutter_command.dart:512 )
  | at <asynchronous gap> | (async )
  | at FlutterCommand.findTargetDevice | (flutter_command.dart:542 )
  | at <asynchronous gap> | (async )
  | at AttachCommand.validateCommand | (attach.dart:139 )
  | at <asynchronous gap> | (async )
  | at FlutterCommand.verifyThenRunCommand | (flutter_command.dart:462 )
  | at <asynchronous gap> | (async )
  | at FlutterCommand.run.<anonymous closure> | (flutter_command.dart:390 )
  | at <asynchronous gap> | (async )
  | at AppContext.run.<anonymous closure> | (context.dart:154 )
  | at <asynchronous gap> | (async )
  | at _rootRun | (zone.dart:1124 )
  | at _CustomZone.run | (zone.dart:1021 )
  | at _runZoned | (zone.dart:1516 )
  | at runZoned | (zone.dart:1463 )
  | at AppContext.run | (context.dart:153 )
  | at <asynchronous gap> | (async )
  | at FlutterCommand.run | (flutter_command.dart:381 )
  | at CommandRunner.runCommand | (command_runner.dart:197 )
  | at <asynchronous gap> | (async )
  | at FlutterCommandRunner.runCommand.<anonymous closure> | (flutter_command_runner.dart:396 )
  | at <asynchronous gap> | (async )
  | at AppContext.run.<anonymous closure> | (context.dart:154 )
  | at <asynchronous gap> | (async )
  | at _rootRun | (zone.dart:1124 )
  | at _CustomZone.run | (zone.dart:1021 )
  | at _runZoned | (zone.dart:1516 )
  | at runZoned | (zone.dart:1463 )
  | at AppContext.run | (context.dart:153 )
  | at <asynchronous gap> | (async )
  | at FlutterCommandRunner.runCommand | (flutter_command_runner.dart:356 )
  | at <asynchronous gap> | (async )
  | at CommandRunner.run.<anonymous closure> | (command_runner.dart:112 )
  | at new Future.sync | (future.dart:224 )
  | at CommandRunner.run | (command_runner.dart:112 )
  | at FlutterCommandRunner.run | (flutter_command_runner.dart:242 )
  | at run.<anonymous closure>.<anonymous closure> | (runner.dart:62 )
  | at <asynchronous gap> | (async )
  | at _rootRun | (zone.dart:1124 )
  | at _CustomZone.run | (zone.dart:1021 )
  | at _runZoned | (zone.dart:1516 )
  | at runZoned | (zone.dart:1500 )
  | at run.<anonymous closure> | (runner.dart:60 )
  | at <asynchronous gap> | (async )
  | at AppContext.run.<anonymous closure> | (context.dart:154 )
  | at <asynchronous gap> | (async )
  | at _rootRun | (zone.dart:1124 )
  | at _CustomZone.run | (zone.dart:1021 )
  | at _runZoned | (zone.dart:1516 )
  | at runZoned | (zone.dart:1463 )
  | at AppContext.run | (context.dart:153 )
  | at <asynchronous gap> | (async )
  | at runInContext | (context_runner.dart:58 )
  | at <asynchronous gap> | (async )
  | at run | (runner.dart:51 )
  | at main | (executable.dart:63 )
  | at <asynchronous gap> | (async )
  | at main | (flutter_tools.dart:8 )
  | at _startIsolate.<anonymous closure> | (isolate_patch.dart:303 )
  | at _RawReceivePortImpl._handleMessage | (isolate_patch.dart:172 )

```

- cleanup runSync -> run

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/simulators.dart [hunk 1]
```diff
@@ -83,20 +84,26 @@ class SimControl {
...
     if (results.exitCode != 0) {
       printError('Error executing simctl: ${results.exitCode}\n${results.stderr}');
       return <String, Map<String, dynamic>>{};
     }
-
-    return json.decode(results.stdout)[section.name];
+    try {
+      return json.decode(results.stdout)[section.name];
+    } on FormatException {
+      // We failed to parse the simctl output, or it returned junk.
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/simulators.dart]: Maybe put some of this in a printTrace() or printError() depending on whether we can explain how to download the missing artifacts. Also if there are artifacts missing, could that be detected with a doctor validator?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: I don't think its our missing artifacts, but some sort of internal initialization that isn't done yet. Like, call me again I'm not ready? Just a theory, I didn't find anything searching for the string

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/ios/simulators.dart]: Updated to include printError and raw simctl output

**Review by zanderso — APPROVED** [review]

**Comment by tvolkert** [comment]
It looks like this caused a regression in the `*_ios32_transition_perf` benchmarks, which is hard to believe.  Perhaps the device is going bad?

@digiter

**Pull request merged**